pydantic_ai_slim/pydantic_ai/models/anthropic.py

```diff
@@ -663,26 +652,91 @@ def _add_builtin_tools(
                 )
         return tools, mcp_servers, beta_features
 
-    def _infer_tool_choice(
+    def _infer_tool_choice(  # noqa: C901
         self,
-        tools: list[BetaToolUnionParam],
         model_settings: AnthropicModelSettings,
         model_request_parameters: ModelRequestParameters,
-    ) -> BetaToolChoiceParam | None:
-        if not tools:
-            return None
+    ) -> tuple[list[BetaToolUnionParam], BetaToolChoiceParam | None]:
+        """Determine which tools to send and the API tool_choice value.
+
+        Returns:
+            A tuple of (filtered_tools, tool_choice).
+        """
+        thinking_enabled = model_settings.get('anthropic_thinking') is not None
+        function_tools = model_request_parameters.function_tools
+        output_tools = model_request_parameters.output_tools
+
+        resolved = _resolve_tool_choice(model_settings, model_request_parameters)
+
+        if resolved is None:
+            tool_defs_to_send = [*function_tools, *output_tools]
         else:
-            tool_choice: BetaToolChoiceParam
+            tool_defs_to_send = resolved.filter_tools(function_tools, output_tools)
+
+        # Map ToolDefinitions to Anthropic format
+        tools: list[BetaToolUnionParam] = [self._map_tool_definition(t) for t in tool_defs_to_send]
+
+        # Add cache_control to the last tool if enabled
+        if tools and (cache_tool_defs := model_settings.get('anthropic_cache_tool_definitions')):
+            ttl: Literal['5m', '1h'] = '5m' if cache_tool_defs is True else cache_tool_defs
+            last_tool = tools[-1]
+            last_tool['cache_control'] = self._build_cache_control(ttl)
+
+        if not tools:
+            return tools, None
 
+        tool_choice: BetaToolChoiceParam
+
+        if resolved is None:
             if not model_request_parameters.allow_text_output:
                 tool_choice = {'type': 'any'}
             else:
                 tool_choice = {'type': 'auto'}
 
-            if 'parallel_tool_calls' in model_settings:
-                tool_choice['disable_parallel_tool_use'] = not model_settings['parallel_tool_calls']
+        elif resolved.mode == 'auto':
+            if not model_request_parameters.allow_text_output:
+                tool_choice = {'type': 'any'}
+            else:
+                tool_choice = {'type': 'auto'}
+
+        elif resolved.mode == 'required':
+            if thinking_enabled:
+                raise UserError(
+                    "tool_choice='required' is not supported with Anthropic thinking mode. "
+                    'Use `output_type=NativeOutput(...)` or `PromptedOutput(...)` instead.'
+                )
+            tool_choice = {'type': 'any'}
+
+        elif resolved.mode == 'none':
+            if len(output_tools) == 1:
+                tool_choice = {'type': 'tool', 'name': output_tools[0].name}
+            else:
```

Comment:
What if there are 0 output tools? Then I think we should send `{type: 'none'}`, so that the model is forced to generate text. And we should `assert model_request_parameters.allow_text_output` just to be sure (I don't think it's possible to get into a situation with 0 output tools and NOT allowing text)